### ISSUE TITLE
Nesting Objects and Arrays

### ISSUE THREAD

**Issue by tonyfast** [newsdev/archieml.org#4]
ArchieML seems really promising.  I know the project just got started, but I was excited to hack on Archie.  I made this [bl.ock](http://bl.ocks.org/tonyfast/091cbd9d9b134d063630) to test the javascript library.  

In the example, I wanted to convert [`flare.json`](https://gist.githubusercontent.com/mbostock/1093025/raw/a05a94858375bd0ae023f6950a2b13fac5127637/flare.json) into a tree structure, but AchieML doesn't appear to work with nested arrays.  But, it appears as if that is not possible in the current version.  Is this something y'all could imagine including?

**Comment by abstrctn** [comment]
Hi Tony - we wrote off nested arrays in the first version mainly because 1) introducing a nested structure to the syntax makes parsing harder, and 2) the use cases at NYT didn't have the need for that much nesting (encouraging a simple document structure can be good both for writers and developers).

Of course, that was for our use cases; I'd love to discuss how to include nesting where it makes sense.

As an example, take the following way that we could implement it:

```
[array]
  name: first element
  [children]
    child: 1
    child: 2
  []

  name: second element
  [children]
    child: 1
    child: 2
  []
[]
```

The above example essentially re-introduces mandatory closing tags (`[]`) to keep track of what level we're at. In a way, this forces users to make a mental map of the output object in order to understand the structure.

We've found it can be easier for non-programmers to write in more a declarative manner, where the special commands act more like headings.

```
[array1]
...
[array2]
...
{}
global: scope
```

We also want it to be easy for developers to access any value (i.e., construct a jsonpath to get to it) without having to traverse the tree structure of the document (which could be pages long) and count through multiple levels of indices to get `children[9].children[5].children[5].children[12].size`. ArchieML isn't perfect about this at the moment (you still might have to count the index within one array), but one level's much easier than an arbitrary number.

TL;DR, I think making the _syntax_ nested is a direction we wouldn't want to move in.

On the other hand, already allow single-level arrays nested within a larger document using dot notation:

```
[scope.nested]
key: value
```

=> `{scope: {nested: "value"}}`

While it's not part of the existing spec, it would at least be possible to allow further nesting like this, without sacrificing either of the above considerations:

```
[scope.array[0].nested]
key: value
```

=> `{scope: {array: [{nested: {key: "value"}}]}}`

Not that that's very much fun to write...

**Comment by tonyfast** [comment]
Yea it is really clear that the simple document structure is the goal.  Which is fantastic.  I like I how some the parser.  Archie certainly lets non-experts share structured data with that.  

I am asking this questions for two reasons (1) I was curious the direction that this project was going to go and (2) I am very sympathetic to scientists sharing data and gleaning it from research papers.

I think that closing the array is the only solution that makes sense in the scope of the language.  It doesn't seem like a hard question for a non-expert to answer, "When do I stop counting?"

I feel like special commands could be supported with JSON-LD.  I think the minimalism of the archie syntax, adding more may change the simple document structure. Domain specific Archie-flavored Markup languages perhaps though?

Thanks for your response Michael.

**Comment by tonyfast** [comment]
I was talking with @bollwyvl today.  He had suggested nesting arrays as 

```
[foo]
* [.bar]
  * a
  * b
*[.baz]
  * c 
  * d
```

basically the `.` would reference the parent.  Currently, a leading dot creates an empty key which doesn't seem as purposeful.

**Comment by abstrctn** [comment]
I'm excited to see where you think AML could be useful outside of the reporting perspective we've been viewing this from so far, so obviously I can't make any promises, but I'm all for you / @bollwyvl running with ideas like this.

If we're going to introduce this behavior into AML, there are a few things we need to continue to adhere to:
- Object / Array lines (`{scope}`, `{}`, `[array]` and `[]`) should always bring you back to the same scope / position, regardless of any other nesting structure you're currently in. e.g., the lines `{} \n key: value` should always create `{key: value}` at the top level of the document.
- Make sure that any additional syntax we add doesn't make the original syntax more error prone. e.g., new commands have to be very deliberate, and the way that the language "removers" from bad syntax, as mentioned in the first bullet, needs to remain.

The example you mention above seems to work on the first point, but I worry about introducing "impossible" array strings. It's obviously unlikely, but in that example, you could no longer create the simple array:

```
{
  "foo": [
    "[.bar]"
  ]
}
```

Plus, I think we'd start to run into some ambiguities with the parser distinguishing between nested array declarations like `* [.bar]` and inline comments like `* [COMMENT] value`.

We're essentially talking about a multi-character command here, `* [ .`, that tells the parser to start a nested array element. I wonder if there's some other single character we can use that would be a bit more intuitive, and stay away from inline comment conflicts?

Frustrating, because the existing spec is so close to nested arrays, but so far away!

**Comment by bollwyvl** [comment]
Maybe everything is reporting, we just don't know it yet?

Hm, i don't know that if i like the bullet in front of the nested array: thought we were talking about...

``` yaml
[Slides]
Layout: title
Title: baron

Layout: bullets
[.bullets]
* zap
* zoom
[]

Layout: credits
```

Right now the orig js makes:

``` json
{ "Slides": [ { "Layout": "title", "Title": "baron" }, { "Layout": "bullets" } ], "": { "bullets": [ "zap", "zoom" ] }, "Layout": "credits" }
```

But i think you know what i meant.

I think empty keys are probably never useful, so is that behavior worth protecting?

However, maybe `^[.bullets]` could push one level onto the context stack, and `^[.]` pop one off...

**Comment by bollwyvl** [comment]
And i don't mean literal carets :P

**Comment by bollwyvl** [comment]
Oh, and you can already use numeral keys to achieve what i meant above:

```
[Slides]
Layout: title
Title: baron

Layout: bullets

Layout: credits

[Slides.1.bullets]
* zap
* zoom
[]
```

So dot notation would just remove knowing the patent keys

**Comment by archietse** [comment]
This is definitely an issue that we are running into as we use the spec in more and more projects. Here's another possibility for noting nested arrays/objects:

```
[sections]

id: A
text: Text for section A.
[[subsections]]
id: 1
text: First subsection of section A.
id: 2
text: Second subsection of section A.
[[listItems]]
* First list item of subsection 2, section A.
* Second list item of subsection 2, section A.
[[]] this text should be ignored -- this closes listItems
[[]] end of subsections

id: B
text: Text for section B.
[[nestedList]]
* First list item of nestedList in section B.
* Second list item of nestedList in section B.

[anotherArray] this starts a new array and ends the the sections array and closes out the nestedList in section B.
```

Essentially the rules would be that double brackets indicate that you are starting a nested object/array within an existing object/array instead of jumping back to the root level. You should be able to nest indefinitely by using double brackets.

Empty double brackets close out the immediate nested object -- but it is not necessary to close out the last nested object if you start another top-level object, that is, any single bracketed object.

This needs to be fleshed out further, but any thoughts?

**Comment by tonyfast** [comment]
This is a nice idea too.

Nested arrays are necessary. using the empty object keys and iterating over them with d3.entries which seems to do the trick.

I think it is important that the syntax has a canonical meaning to the unstructured data.  I like that single bracket says that is this a collection of items and then double brackets are part of the collection.  It is easy to think of a double bracket being a collection in a collection of single brackets.  Nesting arrays in arrays seems like more of an arbitrary programmatic concept.  

I have been hacking with Archie a bunch and I think that closing objects is a really nice practice to promote for deep documents.

**Comment by abstrctn** [comment]
Thanks for your suggestions everyone! Sorry for the delay in getting back on this - this is a topic a lot of people care about, and I'd love to reach a consensus to get this in for the 1.0 release of the spec.

Obviously speak up if I'm getting this wrong, but I think we generally agree that closing nested arrays will be necessary. But to avoid requiring it for top-level arrays as well, we'll need a command that isn't `[]`.

I think the example that @archietse gave gets very close. The principles that it adds to the spec can be summed up with:
- Add an additional command to begin an array within the current array element.
- And a matching command to close that new array

As long as those two commands are distinguishable from existing commands, I don't think we'll run into any conflicts... Closing commands would always be necessary before continuing to define arrays or keys within a nested element, but at any point you can return to the global scope with `[]` or `{}`.

As for what the commands are, I'm concerned with the leading dot notation (`[.sublevel]`) and the double brackets (`[[sublevel]]`), since they both are very similar to commands that exist in some form already. (Single brackets for both arrays and inline comments already make me uncomfortable; I'd like to avoid adding more.)

I played around with using `:end` to end the current array, but unfortunately that would cause ambiguity as to whether you're closing the array, or the most recent key. But I like having the "close array" key be more similar to `:end` than to `[]`, since I want to encourage people to think of `[]` and `{}` almost as "return to home".

Here's a variation on @archietse's example that uses `[sublevel]:` and `:[end]` as the two new commands, which are meant to make the nested array appear similar in syntax to string/value pairs. `[sublevel]:` will wind up in the same place in the output as a string `sublevel:` key would, it's just going to be of type array, and I think this syntax conveys that similarity pretty clearly.

```
[sections]

id: A
text: Text for section A.
[subsections]:
  id: 1
  text: First subsection of section A.

  id: 2
  text: Second subsection of section A.

  [sub1]:
    * First list item of subsection 2, section A.
    * Second list item of subsection 2, section A.
  :[end]

  [sub2]:
    first: 1st
    second: 2nd

    first: 1st
    second: 2nd
    [deepest]:
      * Very Deep Value
    :[end]
  :[end]

  id: 3
  text: Third subsection of section A.
        This subsection has multiple lines.
        :end

[] Back to the global scope
toplevel: value
```

I wish requiring the closing arrays weren't necessary, mainly because a current strength of ArchieML is its recoverability. Any errors you accidentally introduce tend to be corrected when you write another `[array]` or `{object}`. And while that's still the case, it isn't the case within a nested array element. But I think I'm expecting too much :).

**Comment by archietse** [comment]
@mericson and I were brainstorming and liked the dot notation suggested by @tonyfast because it is  clean. Here's one way we think it could work:

```
[sections]

id: A
text: Text for section A.
    [.subsections]
    id: 1
    text: First subsection of section A.
    id: 2
    text: Second subsection of section A.
        [.listItems]
        * First list item of subsection 2, section A.
        * Second list item of subsection 2, section A.
        [] this should close sections.A.subsection.2.listItems
    [] this should close sections.A.subsection.2

id: B  [this initiates a new element in the sections array]
text: Text for section B.
```

In this schema, a preceding dot means to nest that object under the current parent. "[]" would always mean to close the immediate parent object that is not already closed.

This would mean that "[]" no longer means return to root, but it does have a meaning that is consistent in all contexts whether it is closing a root-level object or a nested one.

**Comment by tonyfast** [comment]
I really like this syntax.  It is really close to the way CSS selectors access classes; the syntax is close to a standard.

What would happen if you didn't use the first closing bracket

```
[sections]

id: A
text: Text for section A.
    [.subsections]
    id: 1
    text: First subsection of section A.
    id: 2
    text: Second subsection of section A.
        [.listItems]
        * First list item of subsection 2, section A.
        * Second list item of subsection 2, section A.
    [] this should close sections.A.subsection.2.listItems

id: B  [this initiates a new element in the sections array]
text: Text for section B.
```

Where will `sections.A.subsection.2` close?

**Comment by archietse** [comment]
In your example, without the first closing bracket, the line `id: B` initiates a new array element under `.subsections`, which closes `sections.A.subsection.2` since `id` is the `.subsections` key that indicates a new array element.

(Note that in my example, the indenting is only for clarifying the data structure, but that indenting is ignored in the spec.)

If we ultimately allow nesting, I think we would generally want to encourage the use of closing brackets whenever you use nested elements -- but I think that there are situations where we could make it easier to type in data by automatically closing elements when possible. For example if it saw this ...

```
[sections]

id: A
text: Text for section A.
    [.subsections]
    id: 1
    text: First subsection of section A.
    id: 2
    text: Second subsection of section A.
        [.listItems]
        * First list item of subsection 2, section A.
        * Second list item of subsection 2, section A.
    id: 3 [this automatically closes .listItems since it is not a valid simple array element]
    text: Third subsection of section A.
    [] this should close sections.A.subsection.2

id: B  [this initiates a new element in the sections array]
text: Text for section B.
```

... the parser would automatically close `.listItems` in `subsection.2` because the line after the `second list item` was not a valid list item, but it was a valid key. Because it is not just a valid key, but the initiating key, it initiates a new object in `subsections`. If it is any other valid key, it would be just another `key: value` pair in `subsection.2`.

Of course it would also be valid to manually close `.listItems` by using a `[]`.

**Comment by tonyfast** [comment]
That is not a critical breaking feature really.  It gets most of the data the right.

The idea of closing an array should be simple enough for most people to understand when they stop counting.

It is hard to discuss the impact of a poorly typed syntax without seeing what it is being applied to.  I imagine that Archie is an intermediate representation of data used by Javascript that will convert the data to dynamic HTML presentations.  A partially broken serialization should not completely break a [web-based representation](https://github.com/newsdev/archieml.org/issues/1#issuecomment-87428703).

**Comment by abstrctn** [comment]
I like this - thanks for the suggestion @tonyfast and the examples @archietse.

"Automatic closing" brings up a parsing question: The spec defines arrays as either "complex" or "simple", where items are either objects or strings, determined by whether the parser encounters a `key: value` or a `* bullet` line first.

It also defines the following points, which I don't believe we've discussed as a group, so each of which I think is open for discussion:
- When inside a bullet array, `key: value` lines are treated as unstructured text and ignored, and vice versa.
- Arrays of either type can be reopened, and they retain their prior designation as "complex" or "simple."

If we include automatic array closing, then (for example) a `key: value` line inside a simple array would be semantically significant, equivalent to `[] \n key:value`. It seems that `* bullets` inside complex arrays should continue to be ignored, since you can never move up a level into a simple array.

**Comment by archietse** [comment]
After some further discussion, it is probably best if we require nested objects to always be manually closed, at least in the initial implementation of nested objects. This will almost certainly always be allowed, so documents using this syntax will not break in the future.

There appear to be some logical situations where auto-closing of nested objects makes sense, but it's probably better if we try out nesting with manual closing for a while so we can get a better feel for how it works in practice before making up the rules for auto-closing.

**Comment by tonyfast** [comment]
+1

**Comment by abstrctn** [comment]
Thanks @archietse, that makes sense. I'm gonna go ahead and work on a branch of archieml-js to test this out, based on the example in Archie's comment [here](https://github.com/newsdev/archieml.org/issues/4#issuecomment-87358161). The one difference (for now) will be that both complex AND simple arrays that are nested will require closing `[]`s.

**Comment by abstrctn** [comment]
Implemented a version of this for people to play around with in newsdev/archieml-js#12.

**Comment by abstrctn** [comment]
Anyone have any objections to the behavior as implemented in newsdev/archieml-js#12? @archietse, have any frustrations come up in your use so far?

**Comment by russellsamora** [comment]
Using right now in a project. Working as expected.

**Comment by archietse** [comment]
We've used it on a number of breaking news projects and everything appears to be working well.
